extractor/controller.scala

```diff
@@ -0,0 +1,113 @@
+package morpheus
+package extractors
+
+import scala.meta._
+import scala.meta.dialects.Scala211
+
+package object controller {
+
+  private[this] def extractMethod(m: Decl.Def): String =
+    m.mods.collectFirst {
+      case Mod.Annot(Ctor.Ref.Name("query")) => "get"
+      case Mod.Annot(Ctor.Ref.Name("command")) => "post"
+    }.get
+
+  private[this] def extractAuthenticated(m: Decl.Def): Boolean =
+    m.mods.collectFirst {
+      case Mod.Annot(Ctor.Ref.Name("token")) => ()
+    }.nonEmpty
+
+  private[this] def extractParams(m: Decl.Def, paramsDesc: List[ParamDesc], inBody: Boolean): List[intermediate.RouteParam] = {
+    m.paramss.headOption.map { params => params.map { p =>
+
+      val (tpe, required) = p.decltpe.collectFirst {
+        case Type.Apply(Type.Name("Option"), Seq(t)) => (tpeToIntermediate(t), false)
+        case t: Type => (tpeToIntermediate(t), true)
+      }.head
+
+      val name = p.name.syntax
+      intermediate.RouteParam(
+        name = Option(p.name.syntax),
+        tpe = tpe,
+        required = required,
+        desc = paramsDesc.find(_.name == name).flatMap(_.desc),
+        inBody = inBody
+      )
+    }}.getOrElse(Nil).toList
+  }
+
+  private[this] def extractReturnType(m: Decl.Def): intermediate.Type =
+    m.decltpe.collect {
+      case Type.Apply(
+        Type.Name("Future"),
+        Seq(
+          Type.Apply(
+            Type.Name("Either"),
+            Seq(_, tpe)))) => tpeToIntermediate(tpe)
+    }.headOption.getOrElse {
+        throw new Exception(s"""
+          |This method misses an explicit return type
+          |
+          |  ${m.syntax}
+          |
+          |It should be in the form of Future[Either[A, B]].
+        """.stripMargin)
+      }
+
+  def extractAllRoutes(source: Source): List[intermediate.Route] =
+    source.collect { case t: Defn.Trait => t }.flatMap(t => extractRoute(source, t))
+
+  def extractRoute(source: Source, t: Defn.Trait): List[intermediate.Route] = {
+    val methods = t.collect {
+      case m: Decl.Def if m.mods.collect {
+        case Mod.Annot(Ctor.Ref.Name("query" | "command")) => ()
+      }.nonEmpty => m
+    }
+
+    val (controllerName, name) = t.mods.collectFirst {
+      case Mod.Annot(
+        Term.Apply(
+          Ctor.Ref.Name("path"),
+          Seq(Lit(n: String)))) => (t.name.value, n)
+    }.getOrElse(t.name.value, t.name.value)
+
+    methods.map { m =>
+      val scaladoc = findRelatedComment(source, m)
+      val (desc, tagsDesc) = extractDescAndTagsFromComment(scaladoc)
```

Comment:
this will ideally be replaced by https://github.com/scalameta/scalameta/issues/556 as soon as its ready